plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/CopyPropertiesAction.java

```diff
@@ -0,0 +1,116 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(CopyPropertiesAction.class);
+    private ProfileService profileService;
+
+    public void setProfileService(ProfileService profileService) {
+        this.profileService = profileService;
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public int execute(Action action, Event event) {
+        boolean changed = false;
+        List<String> mandatoryPropTypeSystemTags = (List<String>) action.getParameterValues().get("mandatoryPropTypeSystemTag");
+        String singleValueStrategy = (String) action.getParameterValues().get("singleValueStrategy");
+        for (Map.Entry<String, Object> entry : getEventPropsToCopy(action, event).entrySet()) {
+            // propType Check
+            PropertyType propertyType = profileService.getPropertyType(entry.getKey());
+            Object previousValue = event.getProfile().getProperty(entry.getKey());
+            if (mandatoryPropTypeSystemTags != null && mandatoryPropTypeSystemTags.size() > 0) {
+                if (propertyType == null || propertyType.getMetadata() == null || propertyType.getMetadata().getSystemTags() == null
+                        || !propertyType.getMetadata().getSystemTags().containsAll(mandatoryPropTypeSystemTags)) {
+                    continue;
+                }
+            }
+            String propertyName = "properties." + entry.getKey();
+
+            if (previousValue == null && propertyType == null) {
+                changed = changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), "alwaysSet");
+            } else if (previousValue != null) {
+                if (previousValue instanceof List) {
+                    changed = changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), "addValues");
+                } else if (entry.getValue() instanceof List) {
+                    logger.error("A single property named {} is already set on the profile. Impossible to replace with a list",
+                            entry.getKey());
+                } else {
+                    changed =
+                            changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), singleValueStrategy);
+                }
+            } else {
+                if (propertyType.isMultivalued()) {
+                    changed = changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), "addValues");
+                } else if (entry.getValue() instanceof List) {
+                    logger.error("The property {} should contains a single value as declared in the property types", entry.getKey());
+                } else {
+                    changed =
+                            changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), singleValueStrategy);
+                }
+            }
+        }
```

Comment:
can be simplified: (also we should avoid logging info from the event to avoid creating log injection backdoor)

```
else {
     boolean multipleIsExpected = previousValue != null ? previousValue instanceof List : propertyType.isMultivalued();

      if (multipleIsExpected) {
            changed = changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), "addValues");
      } else if (entry.getValue() instanceof List) {
            logger.error("Impossible to copy the property of type List to the profile, either a single value already exist on the profile or the property type is declared as a single value property. Enable debug log level for more information");
            if (logger.isDebugEnabled()) {
                  logger.debug("cannot copy property {}, because it's a List", entry.getKey())
            }
      } else {
            changed = changed || PropertyHelper.setProperty(event.getProfile(), propertyName, entry.getValue(), singleValueStrategy);
      }
}
```